src/query/sql/src/planner/optimizer/decorrelate/subquery_rewriter.rs

```diff
@@ -452,6 +460,169 @@ impl SubqueryRewriter {
+                            self.replacer.visit(scalar_expr)?;
+                        }
+                        self.predicates.append(&mut filter.predicates);
+                    }
+                    _ => {
+                        self.failed = true;
+                        return Ok(());
+                    }
+                }
+                Ok(())
+            }
+
+            fn result(self) -> Option<ScalarExpr> {
+                (!self.failed).then(|| {
+                    self.predicates.into_iter().fold(
+                        ScalarExpr::ConstantExpr(ConstantExpr {
+                            span: None,
+                            value: Scalar::Boolean(true),
+                        }),
+                        |acc, expr| {
+                            ScalarExpr::FunctionCall(FunctionCall {
+                                span: None,
+                                func_name: "and".to_string(),
+                                params: vec![],
+                                arguments: vec![acc, expr],
+                            })
+                        },
+                    )
+                })
+            }
+        }
+
+        #[derive(Debug)]
+        struct BindingReplacer {
+            new_column_bindings: HashMap<IndexType, ColumnBinding>,
+        }
+
+        impl VisitorMut<'_> for BindingReplacer {
+            fn visit_bound_column_ref(&mut self, col: &mut BoundColumnRef) -> Result<()> {
+                if let Some(new_binding) = self.new_column_bindings.get(&col.column.index) {
+                    col.column = new_binding.clone();
+                }
+                Ok(())
+            }
+        }
+
+        let right_expr_binding = &subquery.output_column;
+        let (SubqueryType::Any, Some(box ScalarExpr::BoundColumnRef(left_column))) =
+            (&subquery.typ, &subquery.child_expr)
+        else {
+            return Ok(None);
+        };
+        // `xx in null` != `where xx = null`
+        if left_column.column.data_type.is_nullable() || !is_conjunctive_predicate {
+            return Ok(None);
+        }
+        let (Some(left_table_index), Some(right_table_index)) = (
+            left_column.column.table_index,
+            right_expr_binding.table_index,
+        ) else {
+            return Ok(None);
+        };
+        // restore possible aliases or duplicate loaded tables by `source_table_id` to determine whether they are the same columns of the same table
+        let (Some(left_source_binding), Some(right_source_binding)) = (
+            left_column.column.as_source(),
+            right_expr_binding.as_source(),
+        ) else {
+            return Ok(None);
+        };
+        // when left is a join, filtering becomes complicated. If in subquery is eliminated,
+        // the filter will be pushed down to Scan before join, resulting in inaccurate data results.
+        if left.has_join(Some(&JoinType::Inner)) || left_source_binding != right_source_binding {
```

Comment:
Even if the left column and the right column are not the same column, we can optimize them by adding the filter `left_col = right_col`.